packages/flutter/lib/src/widgets/navigator.dart

```diff
@@ -1961,14 +1966,19 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
             observer.didRemove(removedRoute, newPrecedingRoute);
           removedRoute.dispose();
         }
-
-        if (newPrecedingRoute != null)
-          newPrecedingRoute.didChangeNext(newRoute);
+      }
+      final int finalIndex = _history.indexOf(newRoute);
+      if (finalIndex > 0) {
+        newRoute.didChangePrevious(_history[finalIndex - 1]);
       }
     });
 
     // Notify for newRoute
     newRoute.didChangeNext(null);
+    if (precedingRoute != null) {
+      newRoute.didChangePrevious(precedingRoute);
```

Comment:
I don't think so on this one. Even though the `precedingRoute` is being removed from the history, it is still there (and we want to animate it). Similarly, we need to set the `precedingRoute` next to `newRoute` despite `precedingRoute` being removed from the history. 
It would seem odd to me to if `precedingRoute` next must be `newRoute`, but `newRoute` previous is not `precedingRoute`.

Reply:
No one is notifying ```newPrecedingRoute.didChangeNext(newRoute);``` and  ```newRoute.didChangePrevious(newPrecedingRoute)``` in this method, we should make sure we have it here before the transition.

I can't see an immediately harm will do to notify  ```newRoute.didChangePrevious(precedingRoute);``` where we previously didn't. The only part in our code that depends on this is in CupertinoPageRoute where it use this notify to update the label of back button. As long as we call ```newRoute.didChangePrevious(newPrecedingRoute)``` after ```newRoute.didChangePrevious(precedingRoute);```, we should be fine in this case. The reason I brought this up is we will likely have to call this two notify back to back which seems weird to me. However, as you mentioned, it probably make more sense to notify ```newRoute.didChangePrevious(precedingRoute);``` to make thing more consistent. This also apply to pushReplacement, we should make sure pushReplacement and pushAndRemoveUntil are in sync